handler.go

```diff
@@ -0,0 +1,614 @@
+	state := request.Request{W: w, Req: r}
+
+	// Record response to get status code and size of the reply.
+	rw := dnstest.NewRecorder(w)
+
+	// Get the response for the DNS lookup from the plugin chain.
+	status, err := plugin.NextOrFailure(resolver.Name(), resolver.Next, ctx, rw, r)
+
+	// Get the DNS name from the DNS lookup request.
+	qname := strings.ToLower(state.QName())
+
+	var regularDnsInfo, wildcardDnsInfo namespaceDNSInfo
+	var regularDNSExists, wildcardDNSExists bool
+
+	// Check if the query was for a wildcard DNS name or a regular DNS name.
+	if isWildcard(qname) {
+		// Get the wildcard DNS name info, if it exists.
+		wildcardDnsInfo, wildcardDNSExists = resolver.wildcardDNSInfo[qname]
+	} else {
+		// Get the regular DNS name info, if it exists.
+		regularDnsInfo, regularDNSExists = resolver.regularDNSInfo[qname]
+
+		// Get the corresponding wildcard DNS name for the reguar DNS name.
+		wildcard := getWildcard(qname)
+		// Get the wildcard DNS name info, if it exists.
+		wildcardDnsInfo, wildcardDNSExists = resolver.wildcardDNSInfo[wildcard]
+	}
+
+	// If neither regular DNS name info nor wildcard DNS name info exists for the DNS name
+	// then return the response received from the plugin chain.
+	if !regularDNSExists && !wildcardDNSExists {
+		return status, err
+	}
+
+	// Check if the DNS lookup is unsuccessful or an error is encountered during the lookup.
+	if status != dns.RcodeSuccess || err != nil {
+		// WaitGroup variable used to wait for the completion of update of DNSNameResolver CRs
+		// corresponding to the regular and the wildcard DNS names.
+		var wg sync.WaitGroup
+
+		// If regular DNS name info exists then update the corresponding DNSNameResolver CR for
+		// the DNS lookup failure.
+		if regularDNSExists {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				resolver.updateResolvedNamesFailure(ctx, regularDnsInfo, qname, status)
+			}()
+		}
+
+		// If wildcard DNS name info exists then update the corresponding DNSNameResolver CR for
+		// the DNS lookup failure.
+		if wildcardDNSExists {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				resolver.updateResolvedNamesFailure(ctx, wildcardDnsInfo, qname, status)
+			}()
+		}
+
+		// Wait for the goroutines to complete.
+		wg.Wait()
+
+		// Return the response received from the plugin chain.
+		return status, err
+	}
+
+	// Get the IP addresses and the corresponding TTLs in a map. Only A and AAAA type DNS records
+	// are considered.
+	ipTTLs := make(map[string]int32)
+	for _, answer := range rw.Msg.Answer {
+		switch state.QType() {
+		case dns.TypeA:
+			if rec, ok := answer.(*dns.A); ok {
+				ttl := int32(rec.Hdr.Ttl)
+				if ttl == 0 {
+					ttl = resolver.minimumTTL
+				}
+				ipTTLs[rec.A.String()] = ttl
+			}
+		case dns.TypeAAAA:
+			if rec, ok := answer.(*dns.AAAA); ok {
+				ttl := int32(rec.Hdr.Ttl)
+				if ttl == 0 {
+					ttl = resolver.minimumTTL
+				}
+				ipTTLs[rec.AAAA.String()] = ttl
+			}
+		default:
+			return status, err
```

Comment:
Actually all records will be of the same type. So the first element itself will not match A or AAAA record and will be returned.

Reply:
I see, so this `default` covers the case when a name put into `spec.name` corresponds to a record different from A/AAAA type (e.g. NS). What if such a lookup fails? We won't even reach this part of code, will we?

Reply:
The failure will be handled before reaching this part of the code. The `status` will indicate the whether the lookup was successful or not and the `err` will indicate if any error was encountered during the lookup.

Reply:
> The failure will be handled before reaching this part of the code. The status will indicate the whether the lookup was successful or not and the err will indicate if any error was encountered during the lookup.

That seems to be a little inconsistent. If the dns name corresponds to a non A/AAAA record the status of the CR will be updated is the lookup fails but the same status will be left intact if the lookup passes.  Should we report an error condition in case the lookup response doesn't correspond to A/AAAA record?

Reply:
Then we need to check the type of the DNS record before checking the `status`/`err`.

Reply:
If we check the type of DNS record first, then we don't update the status of the CR if the type is not of A/AAAA type. Even if the lookup for the non A/AAAA record failed, the status would not be updated with the error message.

The purpose of the error message was more for debugging lookup failures of A/AAAA records and the reason for the failure.